pkg/maps/ctmap/gc/gc.go

```diff
@@ -148,188 +173,203 @@ func (gc *GC) isFullGC(ipv4, ipv6 bool) bool {
+				return
+			}
+			// mute before draining so that no more wakeups are queued just
+			// after we have drained
+			gc.signalHandler.MuteSignals()
+			triggeredBySignal = true
+			ipv4 = false
+			ipv6 = false
+			if x == SignalProtoV4 {
+				ipv4 = true
+			} else if x == SignalProtoV6 {
+				ipv6 = true
+			}
+			// Drain current queue since we just woke up anyway.
+			for len(gc.signalHandler.Signals()) > 0 {
+				x := <-gc.signalHandler.Signals()
 				if x == SignalProtoV4 {
 					ipv4 = true
 				} else if x == SignalProtoV6 {
 					ipv6 = true
 				}
-				// Drain current queue since we just woke up anyway.
-				for len(gc.signalHandler.Signals()) > 0 {
-					x := <-gc.signalHandler.Signals()
-					if x == SignalProtoV4 {
-						ipv4 = true
-					} else if x == SignalProtoV6 {
-						ipv6 = true
-					}
-				}
-			case <-time.After(interval):
-				gc.signalHandler.MuteSignals()
-				ipv4 = gc.ipv4
-				ipv6 = gc.ipv6
 			}
-		}
-	}()
 
-	// Start a background go routine that waits to see if either the initial scan completes before
-	// our expected time of 30 seconds.
-	// This is to notify users of potential issues affecting initial scan performance.
-	go func() {
-		select {
-		case <-initialScanComplete:
-		case <-stdtime.After(initialGCInterval):
-			gc.logger.Warn("Failed to perform initial ctmap gc scan within expected duration." +
-				"This may be caused by large ctmap sizes or by constraint CPU resources upon start." +
-				"Delayed initial ctmap scan may result in delayed map pressure metrics for ctmap.")
+		case <-ctx.Done():
+			gc.logger.Info("Context cancelled. Stopping conntrack garbage collector")
+			return
+
+		case <-time.After(interval):
+			gc.signalHandler.MuteSignals()
+			ipv4 = gc.ipv4
+			ipv6 = gc.ipv6
 		}
-	}()
-
-	if runMapPressureDaemon {
-		// Wait until after initial scan is complete prior to starting ctmap metrics controller.
-		go func() {
-			<-initialScanComplete
-			gc.logger.Info("Initial scan of connection tracking completed, starting ctmap pressure metrics controller")
-			// Not supporting BPF map pressure for local CT maps as of yet.
-			ctmap.CalculateCTMapPressure(gc.controllerManager, gc.metricsRegistry, ctmap.GlobalMaps(gc.ipv4, gc.ipv6)...)
-		}()
 	}
 }
 
+func (gc *GC) checkInitialScanCompletedInTime(ctx context.Context) error {
+	select {
+	case <-ctx.Done():
+		return nil
+	case <-gc.initialScanComplete:
+		return nil
+	case <-stdtime.After(initialGCInterval):
+		msg := `Failed to perform initial ctmap gc scan within expected duration.
+			"This may be caused by large ctmap sizes or by constraint CPU resources upon start.
+			"Delayed initial ctmap scan may result in delayed map pressure metrics for ctmap.`
+		gc.logger.Warn(msg)
+
+		return errors.New("failed to perform initial ctp map gc scan within expected duration")
+	}
+}
+
+func (gc *GC) initCTMapPressureMetrics() {
+	for _, m := range gc.pressureMaps {
+		m.WithPressureMetric(gc.metricsRegistry)
```

Comment:
WithPressureMetric enables the map pressure using the in-memory bpf map cache, however ctmap is inserted into outside of the agent control plane so the cache size doesn't reflect what's actually in the map.

Reply:
That's why the "pressure maps" are stored in the field `GC.pressureMaps` and later used by `initCTMapPressureMetrics` & `CalculateCTMapPressure`. Therefore they should use the same map instances - and it should be similar to what the current logic is doing (other code paths retrieve their own instances by calling `ctmap.GlobalMaps`).

But maybe i misunderstand you completely - in which case it would be helpful if you could explain this in more detail :)